lib/plugin-api/hydrogen-kernel.js

```diff
@@ -27,6 +77,76 @@ export default class HydrogenKernel {
     }
   }
 
+  /*
+   * The language of the kernel, as specified in its kernelspec
+   */
+  get language(): string {
+    this._assertNotDestroyed();
+    return this._kernel.language;
+  }
+
+  /*
+   * The display name of the kernel, as specified in its kernelspec
+   */
+  get displayName(): string {
+    this._assertNotDestroyed();
+    return this._kernel.displayName;
+  }
+
+  /*
+   * Add a kernel middleware, which allows intercepting and issuing commands to
+   * the kernel.
+   *
+   * All of the middleware object's methods are called with `this` set to the
+   * middleware object. Also note that adding or removing methods from the
+   * middleware object after this method has been called is not supported.
+   *
+   * @param {HydrogenKernelMiddleware} middleware
+   */
+  addMiddleware(middleware: HydrogenKernelMiddleware) {
+    this._assertNotDestroyed();
+
+    const prevMiddlewareStack = this._kernel.middlewareStack;
+    const middlewareStack = Object.assign({}, prevMiddlewareStack);
+
+    // The middleware object we get as an argument may later be mutated to
+    // remove/replace some of its methods. The current plugin API is designed to
+    // ignore any such changes, which requires making a copy here. We use local
+    // variables instead of copying the entire object to make flow happy.
+    if (middleware.interrupt) {
+      const interrupt = middleware.interrupt.bind(middleware);
```

Comment:
Note that `.bind` is new as of this patchset.

If the middleware passed in is a class instance rather than a raw Object, I think it makes sense for `this` to do the right thing. I *think* bind is needed to accomplish this (though I'm not quite familiar enough with javascript to know for sure).

Reply:
I don't think there's a need for this, since all these methods are always invoked using the the dot notation, e.g. `middleware.interrupt()`.

Reply:
Oh! I've just seen what you've defined `const interrupt`. I wouldn't. There is no need and it's an unusual thing to do in JS.

Reply:
Note that I only check `if (middleware.interrupt)` once, not at every request (this is what I mean by "precompiling the middleware stack"). I can't assume that `middleware.interrupt` will continue to exist in the future -- flow will actually warn about this type safety error if I don't store a copy here.

More on precompiling vs. not in a bit -- there are a lot of technical details so it'll take me some time to write them up.